### ISSUE TITLE
Multiple Service Principals Question

### ISSUE THREAD

**Issue by bdschaap** [Azure/aad-pod-identity#1112]
I am having difficulty with AAD Pod Identity with pods that require two identities to authenticate with Azure resources in two AAD tenants.  I need to use AzureIdentity resources with AAD service principals in order to support both AKS and on-prem RKE clusters.  I am using this OSS project to accomplish this rather than the AKS managed add-on.  It works fine for a single service principal.  I have tried a couple approaches but neither worked out.

_Approach 1 - separate app registrations / service principals in each AAD tenant._
Configuration of this required two AzureIdentity resources that shared the same matching selector.  After wiring up the AzureIdentity resources a pod trying to authenticate with an Azure SQL MI resource in one tenant seemed to randomly pick the identity to use so it succeeded about half the time.  Failures also showed up as errors in the nmi pod's logs.  The code for that pod was using the Microsoft.Data.SqlClient library with a connection string that specified to use Managed Identity authentication.  The other pod was authenticating using an app registration that is configured with API permissions via the portal.  That app registration allows client authentication with a service that requires AAD authentication.  Authentication was successful but testing was limited and could possibly have randomly succeeded or failed if I tested further.

_Approach 2 - single app registration shared by a service principal in each AAD tenant_
Configuration of this required one AzureIdentity resource using the auxiliaryTenantIds field.  Authentication succeeded for both the Azure SQL MI resource as well as the client authenticating with a service that requires AAD authentication.  However, each one only worked when the tenantId field was configured with the AAD tenant where the service principal exists.  The auxiliaryTenantIds field didn't seem to have any effect and may not be its intended use case.

Assistance appreciated on how I should proceed.

**Comment by chewong** [comment]
For approach 1, do you know whether there is an option to specify the client ID of the identity when authenticating? If there is no client ID specified, it will use the first identity that matches the selector.

**Comment by bdschaap** [comment]
For approach 1, it seemed to be randomly succeeding or failing when authenticating in order to access the Azure SQL MI resource.  The logs showed it matching to one client ID sometimes but the other client ID other times.  I suspect there isn't any guaranteed order that the identities are matched so it just matches the first one.

For the code associated with the service principal in AAD tenant 1, the Microsoft.Data.SqlClient library provides a UserID field and I tried using the client ID.  However, I later noticed that the docs state that its intended for User Assigned Identities but I am using service principals.  It also says that the object ID of the service principal is should be provided as the value in v2.1 of the library.  The client ID should be provided in v3.0 of the library.  I am currently using v2.1 so I plan to test it again with the object ID.

For code associated with the service principal in AAD tenant 2, the Azure.Identity library's DefaultAzureCredential class provides a ManagedIdentityClientId.  I have not tried it but plan to try it out.

**Comment by bdschaap** [comment]
I tried specifying the object ID of the service principal using Microsoft.Data.SqlClient v2.1 without luck.  I then tried specifying the clientID via the User ID property using Microsoft.Data.SqlClient v3.0 and it worked.  The nmi pod logs state that clientID was provided and the AzureIdentity was matched.  I plan to confirm it works with DefaultAzureCredential ManagedIdentityClientId as well and suspect it will.
